x/arkeo/keeper/msg_server_open_contract.go

```diff
@@ -88,6 +89,9 @@ func (k msgServer) OpenContractValidate(ctx cosmos.Context, msg *types.MsgOpenCo
 		if msg.Rate != provider.PayAsYouGoRate {
 			return errors.Wrapf(types.ErrOpenContractMismatchRate, "pay-as-you-go provider rate is %d, client sent %d", provider.PayAsYouGoRate, msg.Rate)
 		}
+		if msg.SettlementDuration != provider.SettlementDuration {
```

Comment:
Question: Since every provider need to set up their own settlement duration , and also PayAsYouGoRate etc ,   and when user open a contract , the PayAsYouGoRate and SettlementDuration need to be exactly the same as provider set up, so why we ask client to type those in when they open a contract? 
Why not we just fetch it from provider instead?

Reply:
Good question.

This essentially acts like a slippage control in an AMM.  Wether malicious or not, a provider could change their rates, durations, etc while another transaction to open a contract is in the mempool.  This would lead to a user opening a contract with different spec than they were expecting due to the asynchronous nature of the transactions.  

By having the user make these parameters explicit, we ensure they either get exactly what they expected or the transaction reverts.